### PR TITLE
Added caching layer to the backend

### PR DESCRIPTION
**Pull request by AdamAkiva**
Added the following:
1. Caching layer to reduce database load and improve performance.
2. Logs to see whether it works.
3. Handled the thundering herd issue.
4. Added ids to database tables to allow caching without overriding existing values.
5. Changed some stress test parameters to not overload `redis`
6. Changed some callbacks to named functions for better debugging information.
7. Updated packages.
8. Additional minor fixes.
9. Added `redis` related stuff to scripts and docker compose.

### CODE REVIEW

File: scripts/start.sh [hunk 1]
```diff
@@ -12,9 +12,12 @@ BE_DIR="$ROOT_DIR"/be;
 NPM_FE_CACHE_FOLDER="cache/fe";
 NPM_BE_CACHE_FOLDER="cache/be";
 
-DB_DATA_FOLDER="$ROOT_DIR"/db-dev-data;
+DB_DATA_FOLDER="$ROOT_DIR"/db-dev-data/pg;
+CACHE_DATA_FOLDER="$ROOT_DIR"/db-dev-data/redis
 DB_MIGRATIONS_FOLDER="$BE_DIR"/db/migrations;
 
+UV_THREADPOOL_SIZE=$(($(nproc --all) - 1))
```
Comment (Meital-Kozhidov) [inline scripts/start.sh:19]: Care to explain the need? I also do not see the usage, no results other than scripts folder

File: scripts/docker-compose.yml [hunk 2]
```diff
@@ -70,9 +90,13 @@ services:
...
       - ALLOWED_HOSTS=localhost
       - ALLOWED_ORIGINS=*
+      - DB_URL=postgres://admin:admin@dashboard-pg-dev:5432/dashboard-dev
+      - CACHE_URL=redis://dashboard-redis-dev:6379
+      - CACHE_MAX_MEM=512mb
+      - DB_TEST_URL=postgres://admin:admin@dashboard-pg-test:5432/dashboard-test
       - NODE_ENV=development
-      - DB_URI=postgres://admin:admin@dashboard-pg-dev:5432/dashboard-dev
-      - DB_TEST_URI=postgres://admin:admin@dashboard-pg-test:5432/dashboard-test
+      - NODE_NO_WARNINGS=0
```
Comment (Meital-Kozhidov) [inline scripts/docker-compose.yml:98]: Care to explain?

File: be/src/main.ts [hunk 3]
```diff
@@ -13,32 +13,63 @@ import { EventEmitter } from 'node:events';
 // See: https://nodejs.org/api/events.html#capture-rejections-of-promises
 EventEmitter.captureRejections = true;
 
+// The default stack trace limit is 10 calls. Increasing it to a number which
+// we'll never have to think about it again
+Error.stackTraceLimit = 256;
```
Comment (Meital-Kozhidov) [inline be/src/main.ts:18]: This is a comment for me to read about it more (or maybe you have a specific link to provide)

File: be/src/entities/utils.ts [hunk 4]
```diff
@@ -123,26 +123,30 @@ export function maxErr(fieldName: string, maxAmount: number) {
...
   fn: () => T,
-  debug: { instance: ReturnType<typeof Debug>; msg: string }
+  debug: { instance: DebugInstance; msg: string }
 ) {
   const { instance: debugInstance, msg } = debug;
 
-  debugInstance(`Begin --- ${msg}`);
-  const res = fn();
-  debugInstance(`End   --- ${msg}`);
+  try {
```
Comment (Meital-Kozhidov) [inline be/src/entities/utils.ts:130]: why the change? was it unable to throw and now it is?

File: be/src/entities/user/controller.ts [hunk 5]
```diff
@@ -20,14 +20,14 @@ export async function readUsers(
 ) {
   try {
     const queryParams = debugWrapper(
-      () => {
+      function readUsersValidator() {
```
Comment (Meital-Kozhidov) [inline be/src/entities/user/controller.ts:23]: So the change here is to never use arrow syntax, even in these kind of situations?

File: be/src/db/schemas.ts [hunk 6]
```diff
@@ -92,8 +92,11 @@ export const userInfoModel = pgTable(
 export const userCredentialsModel = pgTable(
   'users_credentials',
   {
+    // userId is not a primary key to allow caching without overriding
```
Comment (Meital-Kozhidov) [inline be/src/db/schemas.ts:95]: More information please?

File: be/src/db/schemas.ts [hunk 7]
```diff
@@ -137,15 +143,28 @@ export const userCredentialsModel = pgTable(
...
+      .references(
+        () => {
+          return userInfoModel.id;
+        },
+        { onDelete: 'cascade', onUpdate: 'no action' }
+      ),
+    darkMode: boolean('dark_mode').default(false).notNull(),
+    ...timestamps
+  },
+  (table) => {
```
Comment (Meital-Kozhidov) [inline be/src/db/schemas.ts:163]: I think I understand what's going on, but would love to hear your explanation

File: be/__tests__/config/vite.config.ts [hunk 8]
```diff
@@ -42,8 +42,8 @@ const defaultTestConfig: UserConfig = {
 const stressTestConfig: UserConfig = {
   ...defaultTestConfig,
   fileParallelism: false,
-  testTimeout: 600_000,
-  teardownTimeout: 16_000,
+  testTimeout: 60_000,
```
Comment (Meital-Kozhidov) [inline be/__tests__/config/vite.config.ts:45]: Why the major decrease? (next line too)

File: be/__tests__/utils.ts [hunk 9]
```diff
@@ -199,8 +200,10 @@ export async function stressTest(url: string, options: StressTestOptions) {
       {
         url: url,
         ...options,
-        connections: 2_048,
-        amount: 65_536
```
Comment (Meital-Kozhidov) [inline be/__tests__/utils.ts:203]: What does removing the `amount` mean?

File: be/package.json [hunk 10]
```diff
@@ -5,9 +5,9 @@
   "type": "module",
   "main": "./src/main.ts",
   "scripts": {
-    "start:dev": "NODE_ENV=development DEBUG=dashboard:* npx nodemon --config ./nodemon.json ./src/main.ts",
```
Comment (Meital-Kozhidov) [inline be/package.json:8]: is it `NODE_ENV=development` by default? why did you remove it?

File: be/src/db/cache.ts [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+  }
+
+  public async connect() {
+    await this._handler.connect();
+
+    await this._handler.configSet('maxmemory', this._maxMemory);
+    await this._handler.configSet('maxmemory-policy', 'allkeys-lru');
+  }
+
+  public async get<T>(params: {
```
Comment (Meital-Kozhidov) [inline be/src/db/cache.ts:57]: Maybe separate this function logic to helper functions, since you're doing a few things...

File: be/package.json [hunk 12]
```diff
@@ -5,9 +5,9 @@
...
+    "start:dev": "DEBUG=dashboard:* npx nodemon --config ./nodemon.json ./src/main.ts",
     "test": "NODE_ENV=test npx vitest run --config ./__tests__/config/vite.config.ts",
-    "test:logs": "NODE_ENV=test DEBUG=dashboard:* npx vitest --run --config ./__tests__/config/vite.config.ts",
+    "test:logs": "NODE_ENV=test DEBUG=dashboard:*,-dashboard:database npx vitest --run --config ./__tests__/config/vite.config.ts",
```
Comment (Meital-Kozhidov) [inline be/package.json:10]: Why does this syntax mean?

File: be/src/entities/user/service/read.ts
Comment (Meital-Kozhidov) [inline be/src/entities/user/service/read.ts:1]: 1. Can't we use cache to `readUsers`?
2. why `readUser` is not wrapper inside debug like `readUsers`?

File: scripts/start.sh (hunk 1 above)
Comment (AdamAkiva) [reply scripts/start.sh:19]: Of course.
In short it changes the number of libuv thread pool according to the amount of cores your machine has.
[Longer explanation](https://rabisiddique.medium.com/maximizing-node-js-performance-with-thread-pools-912bacbe529a)

File: scripts/docker-compose.yml (hunk 2 above)
Comment (AdamAkiva) [reply scripts/docker-compose.yml:98]: Yep, no much to say, basically [this](https://nodejs.org/api/cli.html#--no-warnings)
In some version of NODE it was changed to being true by default.
I want to see the any warnings so I disable it

File: be/src/main.ts (hunk 3 above)
Comment (AdamAkiva) [reply be/src/main.ts:18]: 10 functions calls are not a lot consider the abstractions used in some places.
[Node.js docs](https://nodejs.org/api/errors.html#errorstacktracelimit)

File: be/src/entities/utils.ts (hunk 4 above)
Comment (AdamAkiva) [reply be/src/entities/utils.ts:130]: This was changed to this in order to still log the `end` even if the function throw.
Before if the function being called threw, there would not have been any `end` log after the `begin`

File: be/src/entities/user/controller.ts (hunk 5 above)
Comment (AdamAkiva) [reply be/src/entities/user/controller.ts:23]: Reverted in later branches.
The idea was yes, but there is no benefit to naming a function which is wrapper with 3 lines of implementation.
So you can ignore it, it was reverted back to being arrow function(s)

File: be/src/db/schemas.ts (hunk 6 above)
Comment (AdamAkiva) [reply be/src/db/schemas.ts:95]: Sure.
When caching user related data, the key used is a unique id.
When, for example, updating part of a user, credentials in this case, I separated the cache based upon the database tables for ease of use and for just piping the result from the cache back to the user without any need to sanitize the data.
In this way if some fields are updated (since redis saves strings), there is no need to fetch the previous value, sanitize it with the update request and override the previous value.
Since they are separated, we can just override the value

File: be/src/db/schemas.ts (hunk 7 above)
Comment (AdamAkiva) [reply be/src/db/schemas.ts:163]: About the index?
Index in a database is usually used to map values in a hashmap by the database to search for values based upon the index faster.
It is recommended to index fields used by `.where()` queries.
If this was not what you asked, correct me please

File: be/__tests__/config/vite.config.ts (hunk 8 above)
Comment (AdamAkiva) [reply be/__tests__/config/vite.config.ts:45]: There is no need for such a high timeout after I measured the times.
The stress test is based upon time now, 45 seconds to be precise.
Adding to that the 8 seconds of teardown timeout and additional 7 for other cases, 60 seems enough to me

File: be/__tests__/utils.ts (hunk 9 above)
Comment (AdamAkiva) [reply be/__tests__/utils.ts:203]: The stress test is now time based instead of amount.
It makes more sense to me and allows setting a stricter timeout

File: be/package.json (hunk 10 above)
Comment (AdamAkiva) [reply be/package.json:8]: Correct.
The compose files passed it as such, I just used that value.
Another option was to remove the value from the compose file, I preferred this option instead

File: be/src/db/cache.ts [hunk 13]
```diff
@@ -0,0 +1,146 @@
+import {
+  createClient,
+  type DebugInstance,
+  type Logger
+} from '../utils/index.js';
+
+/**********************************************************************************/
+
+export default class Cache {
+  private static readonly ONE_DAY_IN_SECONDS = 86_400;
```
[the remark this answers is not in the document]
Comment (AdamAkiva) [reply be/src/db/cache.ts:10]: Arbitrary value which seemed to make sense to me

File: be/package.json (hunk 12 above)
Comment (AdamAkiva) [reply be/package.json:10]: All logs except the database.
`-` means remove from what was enabled.
See [this](https://github.com/debug-js/debug?tab=readme-ov-file#wildcards)

File: be/src/entities/user/service/read.ts
Comment (AdamAkiva) [reply be/src/entities/user/service/read.ts:1]: Very good question, I though about it a lot and came to the following points while implementing it:
1. The cache may be overpopulated in that way (depends on the amount of users)
2. You would not want to insert user by user because it may take a long long time. However how would you insert an array? Redis saves key-value pairs, what will be the key? It can for example be all keys concatenated, but since users can be requested with filters, how would you know when what the user sent will match what is cached?
3. This is an estimation of mine without checking so take that with a grain of salt. Let's say we save each user to the cache, when fetching it back to the client you would need to make so many cache calls, that I think that one database call should be faster.
4. It's way too much work for little benefit. The get multiple users call should be used very rarely and only by an admin, so I see not much benefit in making it work.
Hopefully these are enough reasons for my choice.
If you disagree let me know and we can discuss it.

File: be/src/db/cache.ts [hunk 14]
```diff
@@ -0,0 +1,146 @@
...
+        // this value in a tuple to match types
+        return [JSON.parse(cachedData)] as T;
+      }
+    } catch (err) {
+      this._logger.error(err, 'Cache read error');
+    }
+
+    const pendingPromise = this._pendingPromises.get(key);
+    if (pendingPromise) {
+      this._debug('Thundering herd prevented');
```
[the remark this answers is not in the document]
Comment (AdamAkiva) [reply be/src/db/cache.ts:84]: Sure, will be added tomorrow or later today

File: be/src/db/schemas.ts (hunk 6 above)
Comment (Meital-Kozhidov) [reply be/src/db/schemas.ts:95]: Lets talk about this tommorrow?

Comment (AdamAkiva) [reply be/src/db/schemas.ts:95]: Sure

**Review by Meital-Kozhidov — APPROVED** [review]
